exporters/etw/include/opentelemetry/exporters/etw/etw_tracer.h

```diff
@@ -0,0 +1,905 @@
+   * @param name
+   * @param timestamp
+   * @return
+   */
+  void AddEvent(nostd::string_view name, core::SystemTimestamp timestamp) noexcept override
+  {
+    owner_.AddEvent(*this, name, timestamp);
+  }
+
+  /**
+   * @brief Add named event with custom timestamp and attributes.
+   * @param name Event name.
+   * @param timestamp Event timestamp.
+   * @param attributes Event attributes.
+   * @return
+   */
+  void AddEvent(nostd::string_view name,
+                core::SystemTimestamp timestamp,
+                const common::KeyValueIterable &attributes) noexcept override
+  {
+    owner_.AddEvent(*this, name, timestamp, attributes);
+  }
+
+  /**
+   * @brief Set Span status
+   * @param code Span status code.
+   * @param description Span description.
+   * @return
+   */
+  void SetStatus(trace::StatusCode code, nostd::string_view description) noexcept override
+  {
+    // TODO: not implemented
+    UNREFERENCED_PARAMETER(code);
+    UNREFERENCED_PARAMETER(description);
+  };
+
+  /**
+   * @brief Sets an attribute on the Span. If the Span previously contained a mapping
+   * for the key, the old value is replaced.
+   *
+   * @param key
+   * @param value
+   * @return
+   */
+  void SetAttribute(nostd::string_view key, const common::AttributeValue &value) noexcept override
+  {
+    // TODO: not implemented
+    UNREFERENCED_PARAMETER(key);
+    UNREFERENCED_PARAMETER(value);
+  };
+
+  /**
+   * @brief Update Span name.
+   *
+   * NOTE: this method is a no-op for streaming implementation.
+   * We cannot change the Span name after it started streaming.
+   *
+   * @param name
+   * @return
+   */
+  void UpdateName(nostd::string_view) noexcept override
+  {
+    // We can't do that!
+    // name_ = name;
+  }
+
+  /**
+   * @brief End Span.
+   * @param EndSpanOptions
+   * @return
+   */
+  void End(const trace::EndSpanOptions &options = {}) noexcept override
+  {
+    if (!has_ended_.exchange(true))
+    {
+      owner_.EndSpan(*this, parent_, options);
+    }
+  }
+
+  /**
+   * @brief Obtain SpanContext
+   * @return
+   */
+  trace::SpanContext GetContext() const noexcept override { return context_; }
+
+  /**
+   * @brief Check if Span is recording data.
+   * @return
+   */
+  bool IsRecording() const noexcept override
+  {
+    // For streaming implementation this should return the state of ETW Listener.
+    // In certain unprivileged environments, ex. containers, it is impossible
+    // to determine if a listener is registered. Thus, we always return true.
+    return true;
+  }
+
+  virtual void SetToken(nostd::unique_ptr<context::Token> &&token) noexcept
+  {
+    // TODO: not implemented
+    UNREFERENCED_PARAMETER(token);
+  }
+
+  /// <summary>
+  /// Get Owner tracer of this Span
+  /// </summary>
+  /// <returns></returns>
+  trace::Tracer &tracer() const noexcept { return this->owner_; };
+};
+
+/**
+ * @brief ETW TracerProvider
+ */
+class TracerProvider : public trace::TracerProvider
```

Comment:
Perhaps name the class as `EtwTracerProvider` to make it differentiable from the abstract base class?

Reply:
Maybe you are right. Maybe not. I had doubts about what is the best name to use. My logical reasoning was that these live in two different namespaces, implement the same behavior and interface. And practically can be used interchangeably. Either concrete class with impl, or base class that defines the interface for the rest of exporter. One lives `trace::` and another lives in `ETW::` namespace. My main doubt is that `ETW::EtwTracerProvider` is a bit repetitive, `ETW` in the namespace, then `Etw` in the actual class name again. From that perspective - I merely follow the logic of what Abseil or GSL is doing. They have `absl::variant` vs `std::variant`, and `gsl::span` instead of `std::span`. Same class name. Same facet. Same behavior. But different namespace. If you don't mind, I'd consider this change for a follow-up PR. We can advise our customers to use a `using` alias if this changes in a follow-up PR.

Reply:
Sure, feel free to revisit this nitpick in future PR.

For the example with `absl::variant` and `std::variant`, I think they are not used in the same place, and there is no inheritance relationship between them, so adding namespace to them is a little redundant. The namespace prefix helps to identify the right class but it is not required as we have `using`.

BTW: very often my ctags is confused to find the right class definition from API or SDK with the same class name.

Reply:
> BTW: very often my ctags is confused to find the right class definition from API or SDK with the same class name.

^ Yes, with this one I agree completely. It makes a tiny bit harder to search via grep. Smarter indexers in Visual Studio IDE and Eclipse are _usually_ okay. But I agree it's easier to grep for unique name.